state/application.go

```diff
@@ -3124,6 +3132,94 @@ func allUnits(st *State, application string) (units []*Unit, err error) {
 	return units, nil
 }
 
+// GetUnitAttachmentInfos returns storage attachment info for units not yet provisioned,
+// based on DesiredScale and UnitCount.
+//
+// This is called by the same worker loop that deploy application && updates scales.
+// So consistency checks can be avoided since provisioning and scale updates are interleaved.
+func (a *Application) GetUnitAttachmentInfos() (unitAttachmentInfos []UnitAttachmentInfo, err error) {
+	// Skips if scaling down or already at desired scale (except initial 1-unit deploy).
+	// Note: Attaching storage is not allowed during deployment if DesiredScale > 1, so that case is ignored.
+	if a.doc.UnitCount >= a.doc.DesiredScale && !(a.doc.UnitCount == 1 && a.doc.DesiredScale == 1) {
+		return unitAttachmentInfos, nil
+	}
+
+	unitIds := []string{}
+
+	start := a.doc.UnitCount
+	if a.doc.UnitCount == 1 && a.doc.DesiredScale == 1 {
+		start = 0
+	}
+	for i := start; i < a.doc.DesiredScale; i++ {
+		unitIds = append(unitIds, strconv.Itoa(i))
+	}
```

Comment:
How about
```
unitIds := []string{"0"}
if !initialDeploy {
  unitIds = []string{}
  for i := a.doc.UnitCount; i < a.doc.DesiredScale; i++ {
    unitIds = append(unitIds, strconv.Ioa(i))
  }
}
```

Maybe also a comment about how the new unit Ids we are interested in follow the behaviour of the statefulset scale up behaviour which allocates new ids in cardinal order.